WIP: vectorised fast{Length,Take,Drop,SplitAt} implementations

axman6 opened the pull request:
Initial work to add (much) faster implementations of Data.Text.{length, take,drop,splitAt} which use SIMD like techniques to process data one `Word#` at a time. It is based on the algorithm described at [Even faster UTF-8 character counting](http://www.daemonology.net/blog/2008-06-05-faster-utf8-strlen.html), but implemented purely in Haskell.

The code assumes it is working with valid utf-8 encoded data.

Implemented functions:
- [x] fastLength
- [x] fastTake
- [x] fastDrop
- [x] fastSplitAt
- [x] fastIndex
- [ ] fastTakeEnd
- [ ] fastDropEnd

Known and potential issues so far:
- [x] 64-bit only, needs 32-bit/generic word size support
- [ ] endianness needs testing on different architectures (hit up erikd for a PPC test eventually)
- [ ] Needs some more eyes to ensure correctness of indexing code
- [ ] This should live somewhere else, which module?

Potential improvements
- [ ] Unroll the inner loop by hand (LLVM should do this, but will the NCG?)
- [ ] use actual SIMD instructions to process 128/256bits per iteration (again, NCG is a problem here)

Bonus content - how does this work?
The algorithm sums the number of bytes which have the pattern `0b10xxxxxx`, i.e., continuation bytes, by checking this for all bytes in the `Word#`, and then adding all those together without the use of branches (the really neat bit is the multiplication of `Word#` by the constant `ones` which causes all the bits representing whether they matched the above pattern to be summed together in the top the number, which is then shifted and added to the number of bytes which are continuations.

Commit:
Add initial fastLength code - WIP

Commit:
Actually test fastLength

Commit:
pull out ByteArray# reference so only offsets are passed in recursive call

Commit:
Make code work on 32 bit and 64 bit.

Commit:
Merge branch 'utf8' into utf8-axman6-fastlength

Commit:
Merge branch 'utf8' into utf8-axman6-fastlength

alexbiehl commented on Data/Text.hs:
```diff
@@ -7,6 +7,10 @@
 {-# LANGUAGE TypeFamilies #-}
 #endif
 
+#if defined(__GLASGOW_HASKELL__) && !defined(__HADDOCK__)
```
Have you tried generating docs with haddock? I am not sure about not including this for haddock. It runs CPP just as ghc. And I imagine it will fail if it can't find `WORD_SIZE_IN_BITS`.

axman6 replied:
Sorry for the late reply - this was copied from elsewhere in the code, I should probably have a look at how this handles haddock at some point too, but there's plenty more other improvements that need to be made (i.e., I don't think this code should live in the module it is, so this will be moved elsewhere too).

Commit:
Factor out the continuation byte counting algorithms.

Commit:
Implement fastNthCodepoint, along with fast{Take,Drop,splitAt}

axman6 commented:
I've added fast implementations of `take`, `drop` and `splitAt`, using a similar algorithm as length. This provides `fastNthCodepoint` which may be useful elsewhere.

I'm also intending to implement a `fastNthCodepointEnd` function for `takeEnd` and `dropEnd` (and maybe `indexEnd` for completeness).

I have added tests for this code, and have run several million tests of each function to ensure they always give identical results to the stream fusion based ones, so I'm quite confident at this point the code is correct (under the assumption the `Text` is valid utf-8 anyway, which I hope is an invariance we maintain when constructing `Text`s)

What I would love input on is where this code should live, and whether it should replace the current implementations of these functions or not. I also haven't been able to get the benchmarks running locally, so would love to see the results (though my initial testing has shown `fastLength` to be consistently 100x faster than `length`, but this shouldn't be surprising since `length` pays the penalty of decoding every character, constructing it, and then throwing it away)

hvr commented:
@axman6 just a thought, would it be easy enough to switch between the optimised ones and the old ones via a manual cabal flag (either by CPP or by conditional `hs-source-dirs:`)? I'll try to make some time later today to look into how to benchmark this.

axman6 commented:
I can't see why not, would be easy with a CPP macro, the argument order is intentionally the same as the standard ones.

hvr commented:
@axman6 ...can we do it then? :-) I think that would make benchmarking a bit easier, as you can simply recompile the benchmark subject with different flags to compare standard <-> simd-ish version

Commit:
Add -fvec-functions flag to cabal to switch to conditionally enable

axman6 commented:
@hvr done, though I haven't updated the tests (it would be good to keep the old versions to have something to run correctness tests against - if we do replace these functions maybe the old ones would move into the tests).

axman6 commented:
One thing to be aware of is that _all_ benchmarks will probably get faster with this, since the majority of tests use `T.length`.

kuk0 commented on Data/Text.hs:
```diff
@@ -660,12 +681,91 @@ fastLength (Text arr (Exts.I# off0#) (Exts.I# len0#)) =
+        | Exts.isTrue# (word2Int# (isContByte# (indexWord8Array# ba off#)))
+            = cleanEnd (off# +# 1#)
+        -- If this is not a contiunuation byte, the previous byte is the last one
+        | otherwise                       = Exts.I# off#
```
this should be (off - off0)  -- a "distance" from the beginning of the text, not an offset
see https://github.com/kuk0/text/commit/f752b218059596251a00b71f3bd75e99124a1ba6

axman6 replied:
Hi @kuk0 I thought this would be more useful in general, if a user of the function needs to compute off-off0 they can but this code avoids having lots of `(off - off0) + off0` throughout the code in the places where the offset is directly needed.

kuk0 reviewed (commented):
this is great work!
here are some benchmarks i did on my laptop:
UTF16 vs. UTF8:
```
benchmarking Pure/take/Text+tiny
time                 46.27 ns   (46.25 ns .. 46.28 ns)
time                 53.96 ns   (53.95 ns .. 53.98 ns)
benchmarking Pure/take/Text+ascii-small
time                 67.34 μs   (67.32 μs .. 67.38 μs)
time                 24.70 μs   (24.69 μs .. 24.71 μs)
benchmarking Pure/take/Text+ascii
time                 59.07 ms   (44.47 ms .. 73.57 ms)
time                 21.97 ms   (21.54 ms .. 22.25 ms)
benchmarking Pure/take/Text+english
time                 3.915 ms   (3.911 ms .. 3.919 ms)
time                 1.429 ms   (1.424 ms .. 1.435 ms)
benchmarking Pure/take/Text+russian
time                 6.277 μs   (6.276 μs .. 6.278 μs)
time                 4.300 μs   (4.294 μs .. 4.308 μs)
benchmarking Pure/take/Text+japanese
time                 4.631 μs   (4.630 μs .. 4.632 μs)
time                 4.607 μs   (4.597 μs .. 4.617 μs)
```

Commit:
Add nthCodepointEnd, for later use in take/dropEnd.

Commit:
Merge branch 'utf8' into utf8-axman6-fastlength

# Conflicts:
#	Data/Text.hs
#	text-utf8.cabal

hvr commented:
btw, it may also be worth to compare with the performance of the equivalent operations in the `text-short` package
